File: trove/backend/internal/service/workouts.go
package service

import (
	"context"
	"fmt"

	"github.com/google/uuid"
	"github.com/saifkazi/trove/backend/internal/dao"
	"github.com/saifkazi/trove/backend/internal/model"
)

type WorkoutService struct {
	store *dao.WorkoutStore
}

func (s *WorkoutService) ListWorkoutTypes(ctx context.Context) ([]model.WorkoutType, error) {
	return s.store.ListTypes(ctx)
}

func (s *WorkoutService) CreateWorkoutType(ctx context.Context, name string, sortOrder int16) (*model.WorkoutType, error) {
	return s.store.CreateType(ctx, name, sortOrder)
}

func (s *WorkoutService) DeleteWorkoutType(ctx context.Context, id string) error {
	uid, err := uuid.Parse(id)
	if err != nil {
		return fmt.Errorf("invalid id: %w", err)
	}
	return s.store.DeleteType(ctx, uid)
}

func (s *WorkoutService) CreateExercise(ctx context.Context, p dao.CreateExerciseParams) (*model.Exercise, error) {
	return s.store.CreateExercise(ctx, p)
}

func (s *WorkoutService) UpdateExercise(ctx context.Context, p dao.UpdateExerciseParams) (*model.Exercise, error) {
	return s.store.UpdateExercise(ctx, p)
}

func (s *WorkoutService) DeleteExercise(ctx context.Context, id string) error {
	uid, err := uuid.Parse(id)
	if err != nil {
		return fmt.Errorf("invalid id: %w", err)
	}
	return s.store.DeleteExercise(ctx, uid)
}

func (s *WorkoutService) ListLogs(ctx context.Context, exerciseID string, weekNumber *int32) ([]model.WorkoutLog, error) {
	uid, err := uuid.Parse(exerciseID)
	if err != nil {
		return nil, fmt.Errorf("invalid exercise_id: %w", err)
	}
	return s.store.ListLogs(ctx, uid, weekNumber)
}

func (s *WorkoutService) UpsertLog(ctx context.Context, p dao.UpsertLogParams) (*model.WorkoutLog, error) {
	return s.store.UpsertLog(ctx, p)
}

func (s *WorkoutService) DeleteLog(ctx context.Context, id string) error {
	uid, err := uuid.Parse(id)
	if err != nil {
		return fmt.Errorf("invalid id: %w", err)
	}
	return s.store.DeleteLog(ctx, uid)
}

File: trove/ui/src/services/workouts.ts
import type { WorkoutType, Exercise, WorkoutLog } from '@/types/api'
import { rpc } from './client'

const SVC = 'trove.v1.workouts.WorkoutService'

export const workoutsApi = {
  listTypes: (): Promise<WorkoutType[]> =>
    rpc<{ workoutTypes?: WorkoutType[] }>(SVC, 'ListWorkoutTypes', {})
      .then(r => r.workoutTypes ?? []),

  createType: (name: string, sortOrder = 0): Promise<WorkoutType> =>
    rpc<{ workoutType: WorkoutType }>(SVC, 'CreateWorkoutType', { name, sortOrder })
      .then(r => r.workoutType),

  deleteType: (id: string): Promise<void> =>
    rpc<Record<string, never>>(SVC, 'DeleteWorkoutType', { id }).then(() => undefined),

  createExercise: (workoutTypeId: string, name: string, sortOrder = 0): Promise<Exercise> =>
    rpc<{ exercise: Exercise }>(SVC, 'CreateExercise', { workoutTypeId, name, sortOrder })
      .then(r => r.exercise),

  deleteExercise: (id: string): Promise<void> =>
    rpc<Record<string, never>>(SVC, 'DeleteExercise', { id }).then(() => undefined),

  listLogs: (exerciseId: string): Promise<WorkoutLog[]> =>
    rpc<{ logs?: WorkoutLog[] }>(SVC, 'ListLogs', { exerciseId })
      .then(r => r.logs ?? []),

  upsertLog: (
    exerciseId: string,
    weekNumber: number,
    sets?: number,
    reps?: string,
    weightKg?: number,
  ): Promise<WorkoutLog> =>
    rpc<{ log: WorkoutLog }>(SVC, 'UpsertLog', { exerciseId, weekNumber, sets, reps, weightKg })
      .then(r => r.log),

  deleteLog: (id: string): Promise<void> =>
    rpc<Record<string, never>>(SVC, 'DeleteLog', { id }).then(() => undefined),
}
